### ISSUE TITLE
Elixir: mix crashes with a SIGSEGV when installing packages.

### ISSUE THREAD

**Issue by rogerbraun** [voidlinux/void-packages#4548]
```
roger@kyouko ~> mix local.hex
fish: “mix local.hex” terminated by signal SIGSEGV (Address boundary error)
```

Tested on two systems, both running x86_64 glibc.

**Comment by zbrdge** [comment]
Was able to reproduce the SIGSEGV in GDB with https://gist.github.com/zbrdge/0d92127aac1968b52fa119acc228dcf0

**Comment by zbrdge** [comment]
Here's a backtrace from that:

(gdb) r
Starting program: /usr/lib/erlang/erts-8.0/bin/erlexec -pa /usr/lib/elixir/bin/../lib/eex/ebin /usr/lib/elixir/bin/../lib/elixir/ebin /usr/lib/elixir/bin/../lib/ex_unit/ebin /usr/lib/elixir/bin/../lib/iex/ebin /usr/lib/elixir/bin/../lib/logger/ebin /usr/lib/elixir/bin/../lib/mix/ebin -elixir ansi_enabled true -noshell -s elixir start_cli -extra /usr/bin/mix local.hex
process 5537 is executing new program: /usr/lib/erlang/erts-8.0/bin/beam.smp  
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
[New Thread 0x7fffb5dff700 (LWP 5541)]
[New Thread 0x7fffb503f700 (LWP 5542)]
[New Thread 0x7fffb5466700 (LWP 5543)]
[New Thread 0x7fffac83e700 (LWP 5544)]  
[New Thread 0x7fffb483e700 (LWP 5545)]
[New Thread 0x7fffb481c700 (LWP 5546)]
[New Thread 0x7fffb47fa700 (LWP 5547)]
[New Thread 0x7fffb47d8700 (LWP 5548)]
[New Thread 0x7fffb47b6700 (LWP 5549)]
[New Thread 0x7fffb4794700 (LWP 5550)]
[New Thread 0x7fffb4772700 (LWP 5551)]
[New Thread 0x7fffb4750700 (LWP 5552)]
[New Thread 0x7fffaffff700 (LWP 5554)]  
[New Thread 0x7fffaf7fe700 (LWP 5555)]
[New Thread 0x7fffaeffd700 (LWP 5556)]
[New Thread 0x7fffae7fc700 (LWP 5557)]
[New Thread 0x7fffadffb700 (LWP 5558)]

Thread 15 "2_scheduler" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffaf7fe700 (LWP 5555)]
0x000055555572703b in enif_make_list_from_array ()
(gdb) bt
#0  0x000055555572703b in enif_make_list_from_array ()
#1  0x00007fffac3fbc9c in ?? () from /usr/lib/erlang/lib/crypto-3.7/priv/lib/crypto.so
#2  0x00005555555a48a7 in process_main ()
#3  0x000055555564b946 in ?? ()
#4  0x00005555557aa266 in ?? ()
#5  0x00007ffff72333f4 in start_thread (arg=0x7fffaf7fe700) at pthread_create.c:333
#6  0x00007ffff6d69bad in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109

**Comment by zbrdge** [comment]
Updated backtrace with debug symbols:

Thread 14 "1_scheduler" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffaffdd700 (LWP 11890)]
0x000055555572703b in enif_make_list_from_array (env=env@entry=0x7fffaffdce30, arr=arr@entry=0x7fffac37fa40 <algo_cipher>, cnt=<optimized out>) at beam/erl_nif.c:1600
1600    beam/erl_nif.c: No such file or directory.
(gdb) bt
#0  0x000055555572703b in enif_make_list_from_array (env=env@entry=0x7fffaffdce30, arr=arr@entry=0x7fffac37fa40 <algo_cipher>, cnt=<optimized out>) at beam/erl_nif.c:1600
#1  0x00007fffac17bc9c in algorithms (env=0x7fffaffdce30, argc=<optimized out>, argv=<optimized out>) at crypto.c:828
#2  0x00005555555a48a7 in process_main () at beam/beam_emu.c:3563
#3  0x000055555564b946 in sched_thread_func (vesdp=0x7fffb5342600) at beam/erl_process.c:8153
#4  0x00005555557aa266 in thr_wrapper (vtwd=0x7fffffffddf0) at pthread/ethread.c:114
#5  0x00007ffff72333f4 in start_thread (arg=0x7fffaffdd700) at pthread_create.c:333
#6  0x00007ffff6d69bad in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109

**Comment by zbrdge** [comment]
Updated *full* backtrace: https://gist.github.com/zbrdge/3283483599df146672fc061942391c42

**Comment by Duncaen** [comment]
This should be fixed by fa89d12c2a5ed3cfd4f0f06bc51b1be25f49eb70

**Comment by rogerbraun** [comment]
Can confirm it fixed the problem!

**Comment by bshaffer** [comment]
Question - why does this throw a segfault on 20 but not on 21?

**Comment by Duncaen** [comment]
The array not dynamically allocated, depending on the openssl/libressl version there are more or less ciphers available at build time.
There are ifdefs for X ciphers, if all are available it there are 22.
With 20 the array is too small and the last one would be written into memory that is not allocated for the array:
https://github.com/Duncaen/otp/blob/4763141dcb61464fdc3ac11273d0a21874670bb5/lib/crypto/c_src/crypto.c#L815

With the libressl version we used it was triggered because all the ciphers where available.

My patch was merged by erlang upstream https://github.com/erlang/otp/pull/1140.
